### DISCUSSION TITLE
Where to get `num_bytes_read` from, when using new `AsyncRead::poll_read`?

### DISCUSSION THREAD

**Discussion by Urhengulas** [tokio-rs/tokio#2990] [Q&A] [answered]
Hello, it's me again 😸 

If one was previously using the `num_bytes_read` of the return of `AsyncRead::poll_read`: Where can one get this information from now, since it isn't returned anymore?

_Actually the docs still says it would be returned, but as far as I get the function signature it is not... ([trait.AsyncRead.poll_read](https://docs.rs/tokio/0.3.0/tokio/io/trait.AsyncRead.html#tymethod.poll_read))._

---

Corresponding code: https://github.com/snapview/tokio-tungstenite/blob/master/src/compat.rs#L142-L160

**Comment by Urhengulas** [comment]
Is it possible to just call [`ReadBud::remaining() -> usize`](https://docs.rs/tokio/0.3.0/tokio/io/struct.ReadBuf.html#method.remaining) before and after the call to `AsyncRead::poll_read`, onto the `ReadBuf` passed to it to get the `num_bytes_read`?

```rust
fn run(a: impl AsyncRead, b: &mut [u8], c: &mut Context) -> usize {
    let mut read_buf = ReadBuf::new(b);
    let before = read_buf.remaining();
    a.poll_read(c, &mut read_buf);
    let num_bytes_read = before - read_buf.remaining();
    num_bytes_read
}
```

Should this work or am I missing something?

**Comment by Darksonn** [comment, marked as answer]
Since you created the `ReadBuf`, you know that it starts with `filled()` empty. Thus, the number of bytes can be obtained with `read_buf.filled().len()` after calling `poll_read`. You may enjoy this snippet:

https://github.com/tokio-rs/tokio/blob/8d17261a4b0e83487f6503816c45376bd82eb41d/tokio-util/src/io/poll_read_buf.rs#L59-L63

**Reply by Urhengulas** [reply]
But isn't the `ReadBuf` already filled, if it is created from a non-empty `&mut [u8]`?

**Reply by Darksonn** [reply]
No, when created from an `&mut [u8]` it will have `filled()` empty. The thing that isn't empty will be `initialized()`.

**Reply by Urhengulas** [reply]
Ahh, I see. Thank you very much :D
